File: framework/languages/java/samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/infrastructure/zlink/spots/handlers/PlayActorPlaceMarkHandler.java
package systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.handlers;

import systems.zlink.framework.CancellationToken;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.framework.handlers.ZLinkSpotActorRequest;
import systems.zlink.framework.spots.ZLinkSpotActorRequestContext;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.actors.PlayActor;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.TicTacToeGame;
import systems.zlink.samples.tictactoe.shared.contracts.PlaceMarkReq;
import systems.zlink.samples.tictactoe.shared.contracts.PlaceMarkRes;

@ZLinkHandlerGroup("play-actor")
public final class PlayActorPlaceMarkHandler {
    @ZLinkSpotActorRequest
    public PlaceMarkRes placeMark(
        TicTacToeGame spot,
        PlayActor actor,
        ZLinkSpotActorRequestContext context,
        PlaceMarkReq request,
        CancellationToken cancellationToken) {
        actor.requireJoinedGame();
        PlaceMarkRes response = spot.placeMark(actor, request.cell());
        return response;
    }
}

File: framework/languages/node/samples/TicTacToe.Ts/Server/Play/Infrastructure/ZLink/Spots/TicTacToeGameSpot/Handlers/play-actor-place-mark-handler.ts
import { Inject, Injectable } from '@nestjs/common';
import {
  ZLINK_SPOT_MANAGER,
  ZLINK_SPOT_OUTBOUND
} from '@zlink-systems/nestjs';
import { ZLinkSpotActorRequest } from '@zlink-systems/framework';
import { SampleNames } from '../../../../../../Configuration/sample-settings';
import type {
  ZLinkSpotActorRequestContext,
  ZLinkSpotManager,
  ZLinkSpotOutbound
} from '@zlink-systems/framework';
import type { PlayActor } from '../../../Actors/play-actor';
import type { PlaceMarkReq, PlaceMarkRes } from '../../../../../../../Shared/Contracts/messages';
import { PlaceMarkAtGameSpotReq } from './tictactoe-game-operation-handlers';

@Injectable()
class PlayActorPlaceMarkHandler {
  constructor(
    @Inject(ZLINK_SPOT_MANAGER) private readonly spotHandles: ZLinkSpotManager,
    @Inject(ZLINK_SPOT_OUTBOUND) private readonly spotOutbound: ZLinkSpotOutbound
  ) {}

  @ZLinkSpotActorRequest('PlaceMarkReq')
  async handle(
    actor: PlayActor,
    _context: ZLinkSpotActorRequestContext,
    request: PlaceMarkReq
  ): Promise<PlaceMarkRes> {
    const spotRid = actor.context.spotRid;
    if (spotRid === undefined) {
      throw new Error(`Actor '${actor.actorId}' is not joined to a game.`);
    }
    const spot = await this.spotHandles.find(spotRid);
    if (spot === undefined) {
      throw new Error(`Game spot '${String(spotRid)}' could not be resolved.`);
    }
    return this.spotOutbound
      .requestToSpot(spot, new PlaceMarkAtGameSpotReq(actor.actorId, request.cell))
      .yield<PlaceMarkRes>();
  }
}

export { PlayActorPlaceMarkHandler };
